File: II-176-3/Harmony.cpp
// BEGIN CUT HERE
#include "precompiled.h"
/*
// END CUT HERE
#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <iostream>
#include <iterator>
#include <map>
#include <numeric>
#include <set>
#include <sstream>
#include <stack>
#include <string>
#include <queue>
#include <vector>
using namespace std;
// BEGIN CUT HERE
*/
#include <Windows.h>
#pragma comment( lib, "Winmm" )
// END CUT HERE
typedef vector<int> vi;
typedef vector<vi> vvi;
#define all(v) (v).begin(), (v).end()
template<typename T> T gcd( T a, T b )
{
	while( a && b ) if ( a > b ) a %= b; else b %= a; return a + b;
}
struct Harmony
{
	struct comp: binary_function<vi, vi, bool>
	{
		vi norm( vi x ) const
		{
			vi result;
			result.push_back( x[0] / gcd( x[0], x[1] ) );
			result.push_back( x[1] / gcd( x[1], x[2] ) );
			result.push_back( x[0] / gcd( x[2], x[0] ) );
			sort( all(result), greater<int>() );
			return result;
		}
		bool operator() ( vi a, vi b ) const
		{
			vi A = norm( a ), B = norm( b );
			if ( A == B )
				return lexicographical_compare( all(a), all(b) );
			return lexicographical_compare( all(A), all(B) );
		}
	};
	vi mostHarmonious( vi frequencies )
	{
		vvi triads;
		int n = frequencies.size();
		sort( all(frequencies) );
		for( int i = 0; i < n; ++i )
			for( int j = i + 1; j < n; ++j )
				for( int k = j + 1; k < n; ++k )
				{
					vi temp;
					temp.push_back( frequencies[i] );
					temp.push_back( frequencies[j] );
					temp.push_back( frequencies[k] );
					triads.push_back( temp );
				}
		return *min_element( all(triads), comp() );
	}
	// BEGIN CUT HERE
	private:
	template <typename T> string print_array( const vector<T>& v ) { ostringstream os; os << "{ "; for( typename vector<T>::const_iterator it = v.begin(); it != v.end(); ++it ) os << '\"' << *it << "\","; os << " }"; return os.str(); }
	template<typename T> void verify_case( int Case, const vector<T>& expected, const vector<T>& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: " << print_array( expected ) << endl; cerr << "\tReceived: " << print_array( received ) << endl; } }
	void verify_case( int Case, const long long& expected, const long long& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const double& expected, const double& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const int& expected, const int& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const string& expected, const string& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	public:
	void run_test( int Case = -1 ) { vector<vi> A0; vector<vi> A1;
	{ /*Test 0*/int _frequencies[] = 
	{200,250,400,320,350}
	; vi frequencies(_frequencies, _frequencies + (sizeof(_frequencies) / sizeof(_frequencies[0]))); A0.push_back( frequencies ); int _result[] = 
	{ 200,  250,  400 }
	; vi result(_result, _result + (sizeof(_result) / sizeof(_result[0]))); A1.push_back( result );  }
	{ /*Test 1*/int _frequencies[] = 
	{440, 320, 750,660, 500,550}
	; vi frequencies(_frequencies, _frequencies + (sizeof(_frequencies) / sizeof(_frequencies[0]))); A0.push_back( frequencies ); int _result[] = 
	{ 440,  550,  660 }
	; vi result(_result, _result + (sizeof(_result) / sizeof(_result[0]))); A1.push_back( result );  }
	{ /*Test 2*/int _frequencies[] = 
	{1960,1000,3050,2341,7253,7864,2000,2352,2940,1534,7234}
	; vi frequencies(_frequencies, _frequencies + (sizeof(_frequencies) / sizeof(_frequencies[0]))); A0.push_back( frequencies ); int _result[] = 
	{ 1960,  2352,  2940 }
	; vi result(_result, _result + (sizeof(_result) / sizeof(_result[0]))); A1.push_back( result );  }
	{ /*Test 3*/int _frequencies[] = 
	{100,200,300,400,500,600,700,800,900,1000}
	; vi frequencies(_frequencies, _frequencies + (sizeof(_frequencies) / sizeof(_frequencies[0]))); A0.push_back( frequencies ); int _result[] = 
	{ 100,  200,  400 }
	; vi result(_result, _result + (sizeof(_result) / sizeof(_result[0]))); A1.push_back( result );  }
	for( int i = 0; i < A0.size(); i++ ) if ( Case == -1 || i == Case ) { DWORD start, finish; start = timeGetTime(); verify_case( i, A1[i], mostHarmonious( A0[i]) ); finish = timeGetTime(); cout << "Elapsed " << finish - start << " msec" << endl; } }
// END CUT HERE
};
// BEGIN CUT HERE
int main()
{
	DWORD start, finish;
	start = timeGetTime();
	Harmony ___test; 
	___test.run_test(); 
	finish = timeGetTime();
	cout << endl << "Elapsed " << finish - start << " msec" << endl;
} 
// END CUT HERE

File: II-176-3/Harmony.cs
using System;
using System.Collections;
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;

public class Harmony 
{
	public int gcd( int a, int b )
	{
		while ( a > 0 && b > 0 )
		{
			if ( a > b ) a %= b;
			else b %= a;
		}
		return a + b;
	}

	public int[] mostHarmonious( int[] frequencies ) 
	{
		Array.Sort( frequencies );
		int n = frequencies.Length;
		int[] result = new int[] {1, 10000, 100000000};
		for( int x = 0; x < n; x++ )
			for( int y = 0; y < x; y++ )
				for( int z = 0; z < y; z++ )
				{
					int[] Y = new int[] {result[1] / gcd( result[0], result[1] ), 
										 result[2] / gcd( result[1], result[2] ), 
										 result[2] / gcd( result[0], result[2] )};
					int[] X = new int[] {frequencies[y] / gcd( frequencies[x], frequencies[y] ), 
										 frequencies[z] / gcd( frequencies[y], frequencies[z] ), 
										 frequencies[z] / gcd( frequencies[z], frequencies[x] )};
					Array.Sort( X );
					Array.Reverse( X );
					Array.Sort( Y );
					Array.Reverse( Y );
					if ( X[0] > Y[0] ) continue;
					if ( X[0] < Y[0] ) 
					{ 
						result = new int[] {frequencies[x], frequencies[y], frequencies[z]};
						continue;
					}
					if ( X[1] > Y[1] ) continue;
					if ( X[1] < Y[1] ) 
					{ 
						result = new int[] {frequencies[x], frequencies[y], frequencies[z]};
						continue;
					}
					if ( X[2] > Y[2] ) continue;
					if ( X[2] < Y[2] ) 
					{ 
						result = new int[] {frequencies[x], frequencies[y], frequencies[z]};
						continue;
					}
				}
		Array.Sort( result );
		return result;
	}

	NumberFormatInfo nfi = new CultureInfo( "en-US", false ).NumberFormat;

// BEGIN CUT HERE
	public static void Main(string[] args)
	{
		DateTime t1 = DateTime.Now;
		eq(1,(new Harmony()).mostHarmonious(new int[] {200,250,400,320,350}),new int[] { 200,  250,  400 });
		eq(2,(new Harmony()).mostHarmonious(new int[] {440, 320, 750,660, 500,550}),new int[] { 440,  550,  660 });
		eq(3,(new Harmony()).mostHarmonious(new int[] {1960,1000,3050,2341,7253,7864,2000,2352,2940,1534,7234}),new int[] { 1960,  2352,  2940 });
		eq(4,(new Harmony()).mostHarmonious(new int[] {100,200,300,400,500,600,700,800,900,1000}),new int[] { 100,  200,  400 });
		DateTime t2 = DateTime.Now;
		TimeSpan t = t2 - t1;
		Console.WriteLine( "{0} sec", t.TotalMilliseconds / 1000 ); 
	}
	private static void eq(int n, int a, int b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, double a, double b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, char a, char b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected '"+b+"', received '"+a+"'.");
	}
	private static void eq(int n, long a, long b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, bool a, bool b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, string a, string b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, int[] a, int[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, double[] a, double[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, long[] a, long[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, string[] a, string[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void print(int a)
	{
		Console.Write(a+" ");
	}
	private static void print(string s)
	{
		Console.Write("\""+s+"\" ");
	}
	private static void print(int[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(double[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(long[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(string[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write("\""+rs[i]+"\"");
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void nl()
	{
		Console.WriteLine();
	}
// END CUT HERE
}
